models/action.go

```diff
@@ -532,32 +532,45 @@ func UpdateIssuesCommit(doer *User, repo *Repository, commits []*PushCommit, bra
 			}
 			refMarked[key] = true
 
-			// only create comments for issues if user has permission for it
-			if perm.IsAdmin() || perm.IsOwner() || perm.CanWrite(UnitTypeIssues) {
-				message := fmt.Sprintf(`<a href="%s/commit/%s">%s</a>`, repo.Link(), c.Sha1, html.EscapeString(c.Message))
-				if err = CreateRefComment(doer, refRepo, refIssue, message, c.Sha1); err != nil {
-					return err
-				}
+			// FIXME: this kind of condition is all over the code, it should be consolidated in a single place
+			canclose := perm.IsAdmin() || perm.IsOwner() || perm.CanWrite(UnitTypeIssues) || refIssue.PosterID == doer.ID
```

Comment:
I think just `canclose := perm.CanWrite(UnitTypeIssues) || refIssue.PosterID == doer.ID` and `cancomment := perm.CanRead(UnitTypeIssues)` is OK.

Reply:
@lunny Done. I've left them in separate lines for clarity, but they can be "merged" in their respective ifs.

Reply:
@lunny It turns out that `IsAdmin()` and `IsOwner()` are not checked by `perm.CanWrite()`/`CanRead()`; those only check the explicit units inside perm. I had to revert the commit. `IsAdmin()` _could_ be left out, but I'd prefer to do that in another PR where we can normalize these functions.